api/v1beta1/auth_config_types.go

```diff
@@ -140,6 +140,10 @@ type Identity struct {
 	// +kubebuilder:default:=0
 	Priority int `json:"priority,omitempty"`
 
+	// Whether this identity config should generate individual observability metrics
+	// +kubebuilder:default:=false
+	Metrics bool `json:"metrics,omitempty"`
```

Comment:
I see the advantages of enabling metrics at the auth config level, the main point being that auth configs that don't need to be monitored can be discarded which will reduce the cost of generating and processing more data but I was just wondering how confusing it can be to monitor these auth configs at a service level as only the auth configs with the metrics enabled will be exposed.

Reply:
> will reduce the cost of generating and processing

I'd also include memory consumption.

I guess I'd rather not have metrics at the AuthConfig level at all, than enabling for all by default. We're talking tons of useless stats.

At the same time, not having those stats can make users' lives very difficult. There are usually simple tweaks that one can do (especially after we improve caching), to improve performance of an AuthConfig, to workaround critical paths momentarily, etc... but only if people are able to know where to act.

I really think we need these metrics and that they have to be opt-in.

> how confusing it can be to monitor these auth configs at a service level

I really don't have a good answer to this, I'm afraid. Maybe if it was simpler to set a separate metrics endpoint per AuthConfig (e.g. /server-metrics/my-authconfig-ns/my-authconfig-name); even so, it wouldn't solve the issue of enabled/disabled.

Do you have any suggestion?

Reply:
There is something similar to this in SSO where each realm has a metrics endpoint. A follow on issue here might be to create a grafana dashboard using these metrics. It would be a practical exercise of trying to use these metrics to visualise useful data?

Reply:
> how confusing it can be to monitor these auth configs at a service level

Introduced new env var `DEEP_METRICS_ENABLED`: enforces deep (evaluator-level) metrics exported for all AuthConfigs, so there is an easier way to set these metrics other than by going AuthConfig by AuthConfig and adding `metrics: true` for each evaluator.

This will be useful for debugging and enabling deep monitoring at the level of service.